### ISSUE TITLE
API: capabilities of df.set_index

### ISSUE THREAD

**Issue by h-vetinari** [pandas-dev/pandas#24046]
This is coming out of a discussion that has stalled #22225 (which is about adding `.set_index` to Series, see #21684). The discussion has shifted away from what capabilities a putative `Series.set_index` should have, but what capabilities `df.set_index` has *currently*.

The main issue (for @jreback) is that `df.set_index` takes arrays:
> > @jreback: There were several attempts to have DataFrame.set_index take an array as well, but these never got off the ground.

> @h-vetinari: I'm not sure *when*, but they certainly did get off the ground:

```
>>> import pandas as pd
>>> import numpy as np
>>> pd.__version__
'0.23.4'
>>>
>>> df = pd.DataFrame(np.random.randint(0, 10, (4, 4)), columns=list('abcd'))
>>> df.set_index(['a',          # label
...               df.index,     # Index
...               df.b ** 2,    # Series
...               df.b.values,  # ndarray
...               list('ABC'), # list
...               'c'])         # label again
              b  d
a   b      c
0 0 0  2 A 1  0  2
8 1 1  4 B 4  1  4
3 2 25 5 C 8  5  5
0 3 9  7 D 2  3  7
```

Further on:
> @jreback: @h-vetinari you are confusing the purpose of `.set_axis`. [...] The problem with `.set_index` on a DataFrame with an array is that it technically can work with an array and not keys. (meaning its not unambiguous)

I don't think I am confusing them. If I want to set the `.index`-attribute of a Series/DataFrame, then using `.set_index` is the most reasonable name by far. If anything, `set_axis` should be a superset of `set_index` (and a putative `set_columns`), that just switches between the two based on the `axis`-kwarg.

More than that, the current capabilities of `df.set_index` are a proper superset of `df.set_axis(axis=0)`**, in that it's possible to fill `keys` with *only* `Series`/`Index`/`ndarray`/`list` etc.:
```
>>> df.set_index(pd.Index(df.a))  # same result as Series directly below
>>> df.set_index(df.a) 
   a  b  c  d
a
0  0  0  1  2
8  8  1  4  4
3  3  5  8  5
0  0  3  2  7
>>> df.set_index(df.a.values)  # same result as list directly below
>>> df.set_index([[0, 8, 3, 0]])
   a  b  c  d
0  0  0  1  2
8  8  1  4  4
3  3  5  8  5
0  0  3  2  7
```
** there is one caveat, in that lists (and only lists; out of all containers) need to be wrapped in another list, i.e. `df.set_index([[0, 8, 3, 0]])` instead of `df.set_index([0, 8, 3, 0])`. This is the heart of the ambiguity that @jreback mentioned above.

Summing up:
* [x] `set_index` is the most natural name for setting the `.index`-attribute
* [x] `df.set_index` *should* be able to process list-likes (as it currently does; this is the source of the ambiguity of the list case).
* [ ] `df.set_axis` should be able to do everything that `df.set_index` does, and just switch between operating on index/columns based on the `axis`-kwarg (after all, `index` and `columns` are the two axes of a DF).
  * [ ] it could be considered to add a method `set_columns` on a `DataFrame` 
  * [ ] The `axis`-kwarg of `set_axis` should just switch between the behaviour of `set_index` (i.e. dealing with keys *and* array-likes) and `set_columns`.
* [ ] `Series.set_index` should support the same signature as `df.set_index`, with the exception of the `drop`-keyword (which only makes sense for column labels).
* [ ] For Series, the `set_index` and `set_axis` methods should be exactly the same.

Since I can't tag @pandas-dev/pandas-core, here are a few individual tags: @jreback @TomAugspurger @jorisvandenbossche @gfyoung @WillAyd @jbrockmendel @jschendel @toobaz.

**Comment by jbrockmendel** [comment]
@h-vetinari should `list('ABC')` in the first example be `list(ABCD')`?  If not then I am confused in several directions.

**Comment by h-vetinari** [comment]
@jbrockmendel 
That was indeed an artefact from merging together several things from the other thread to make this issue...

**Comment by h-vetinari** [comment]
@jreback Any comments here or in #22225?

**Comment by h-vetinari** [comment]
@jreback 
I am honestly stunned by you closing #22225 and then *locking* it after I objected. So much for my motivation to work on some big PRs today. 

> @h-vetinari you are not listening. If you want to raise an issue or comment feel free.

I've opened this issue here for exactly this purpose (discussing your object to *existing* capabilities of `DataFrame.set_index`) over a month ago.

**Comment by toobaz** [comment]
@h-vetinari While I think locking #22225 was an unnecessary move from @jreback , you have to realize that the "''[overruling approving reviews](https://github.com/pandas-dev/pandas/pull/22225)''" thing is not a good argument to raise in such a discussion. True, in pandas we look for devs consensus, but in the end we prefer to do so by argumenting rather than by any form of proper vote. So that comment was not very useful, to use an euphemism.

Back to the topic of the discussion: if I understand correctly, @jreback is argumenting that ``df.set_index`` is already ambiguous enough to make it a bad idea to sponsor its use when passing anything but keys (which would be the only sensible use in ``Series.set_index``); and at the same time, you are suggesting that

>  [ ] For Series, the `set_index` and `set_axis` methods should be exactly the same.

If this summary of the discussion is correct, then I think I am also against introducing ``Series.set_index``.

And actually, I'm probably in favor of deprecating ``df.set_index`` to pass actual values, if ``df.set_axis`` is able to fulfill exactly the same task (didn't check).

I understand you argument about ``df.set_index`` being what one expects to use to set the ``.index`` (to some values)... but ``set_axis`` is simply too long established to be removed, and I find duplication in the API a worse problem than the maybe sub-optimal naming. Or in other words, I think 

>     * [ ]  The `axis`-kwarg of `set_axis` should just switch between the behaviour of `set_index` (i.e. dealing with keys _and_ array-likes) and `set_columns`.

is a bad idea. _''There should be one - and preferably only one - obvious way to do it."_

**Comment by TomAugspurger** [comment]
Agreed that locking was not appropriate.

---

On the issue itself, to me it's pretty clear that `Series.set_index(sequence)` is a limiting case of `DataFrame.set_index(Sequece[sequence])`

```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({"A": [1, 2, 3]})

In [3]: df.set_index([['a', 'b', 'c']])
Out[3]:
   A
a  1
b  2
c  3
```

Since `In[3]` works, I would expect that

```python
In [4]: df.A.set_index([['a', 'b', 'c']])
```

work as well.

**Comment by WillAyd** [comment]
I don't think continuing to post here or in the associated PR is an effective use of anyone's time. Why don't we just add it as a discussion point to the next dev chat?

**Comment by TomAugspurger** [comment]
The implementation is fine, and deserves to go in 0.24.0 if we can agree on
the desired behavior. No need to delay I think.

**Comment by WillAyd** [comment]
There isn't agreeance on desired behavior hence why I suggest moving to a separate forum. I don't think it's something we need to push into 0.24 at the end here either

**Comment by TomAugspurger** [comment]
I really don't think this is a *difficult* decision though, is it? Do we want `Series.set_index` to accept arrays like `DataFrame.set_index` or not? Joris and I are I think +1, or at least ambivalent. Jeff seems to think that DataFrame.set_index doesn't accept arrays (e.g. 
https://github.com/pandas-dev/pandas/pull/22225#issuecomment-451758441)

> `DataFrame.set_index()` which accepts keys (which are column names / levels). NOT an array of values.

which isn't correct, as shown in 
https://github.com/pandas-dev/pandas/issues/24046#issuecomment-451921548.

**Comment by jorisvandenbossche** [comment]
I agree that if it seems difficult to find an agreement, that discussing this on the next dev chat can be more productive / effective. 
But, until now, there hasn't been much discussion on the PR, apart from between Jeff and h-vetinari (other people commented on the PR, but didn't really involve in the API discussion, apart from Tom agreeing with h-vetinari). So I would first still be interested in hearing what other people think about it, as I personally don't really see much reason to object it.

**Comment by toobaz** [comment]
> On the issue itself, to me it's pretty clear that `Series.set_index(sequence)` is a limiting case of `DataFrame.set_index(Sequece[sequence])`

@TomAugspurger We all agree (I think) that it makes sense for ``df.set_index([a_list_of_labels])`` to work. I think @jreback makes a good point however that there is no obvious reason (except parameter ambiguity) for ``df.set_index(a_list_of_labels)`` not to work (since ``df.set_index(Series(a_list_of_labels))`` does), and that this causes a potential confusion that ``df.set_axis`` doesn't. Then maybe we can live with it... but let's admit this is not ideal.

One alternative (which I don't particularly like) is what (I think) ``.groupby(a_list)`` does, i.e., trying to find elements of ``a_list`` in the axis, and fallback to considering them as values otherwise.

**Comment by WillAyd** [comment]
I am -1 due to ambiguity. I don't know what the desired behavior of the following is:

```python
df = pd.DataFrame(np.ones((3,3)))
df.set_index([1, 0, 2])
```

**Comment by toobaz** [comment]
> @toobaz can you show an example of `df._set_index(a_list_of_labels)` vs. `[a_list_of_labels]`? I don't think that #22225 is changing that at all.

No, it's not. But as already stated, if ``df.set_index(values_rather_than_keys)`` is a regrettable legacy causing ambiguity in the API, we'd rather not enhance its usage by paralleling it with ``Series.set_index``, which would do _only_ that (which is already done by ``Series.set_axis``). I actually [suggested](https://github.com/pandas-dev/pandas/issues/24046#issuecomment-451827403) deprecating it... which might not be our final decision, but is certainly related to #22225 .

**Comment by TomAugspurger** [comment]
@toobaz my apologies, I missed the paragraph where you suggested deprecating it. Indeed, if we want to deprecate that then we should go forward in #22225.

**Comment by TomAugspurger** [comment]
On deprecating passing values, rather that column labels to `DataFrame.set_index`: I don't think we should deprecate that. While there is ambiguity, as noted in @WillAyd's example in 
https://github.com/pandas-dev/pandas/issues/24046#issuecomment-451991224, I think it's quite useful to pass a *mix* of labels and keys.

```python
In [11]: df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})

In [12]: df.set_index(["A", [1, 2, 3]])
Out[12]:
     B
A
1 1  4
2 2  5
3 3  6
```

Without that, I think you'd have some ugly


```python
In [18]: df.set_axis(pd.MultiIndex.from_arrays([df.A, [1, 2, 3]]), inplace=False).drop(['A'], axis=1)
Out[18]:
     B
A
1 1  4
2 2  5
3 3  6
```

**Comment by jreback** [comment]
you could just raise / warn on ambguity

**Comment by jreback** [comment]
this needs coupling with possibly deprecating set_axis as well 
because passing values is not documented in any way

**Comment by jorisvandenbossche** [comment]
> this needs coupling with possibly deprecating set_axis as well

I would personally be happy to get rid of `set_axis`, but I think the main problem there is that we don't have an alternative for `df.set_axis(['a', 'b'], axis=1)` (setting column names) ?

**Comment by toobaz** [comment]
> this needs coupling with possibly deprecating set_axis as well

Deprecating the one method that works as expected?!

> I would personally be happy to get rid of `set_axis`, but I think the main problem there is that we don't have an alternative for `df.set_axis(['a', 'b'], axis=1)` (setting column names) ?

The problem is also for ``axis=0``... (unless you know about the nested list trick)

**Comment by toobaz** [comment]
> I think it's quite useful to pass a _mix_ of labels and keys.

I would never do this in my code (I would rather add a column, use ``set_index(append=True)`` and possibly reorder levels). But sure, in this case ``set_index`` is strictly more powerful...

**Comment by h-vetinari** [comment]
> @toobaz: True, in pandas we look for devs consensus, but in the end we prefer to do so **by argumenting** rather than by any form of proper vote. [my emphasis]

That was exactly my point - and locking an on-going discussion is not an argument. But enough of that, thanks for reverting the lock @jreback and the other voices who expressed support.

---------------

The other contributions are quite divergent, so I won't address them one-by-one.

> @jorisvandenbossche: Given the confusion and talking next to each other, it might be good if someone attempts to make a good illustrated and complete summary of the actual discussion.

I attempted as much in the OP (under the assumption that `set_axis` would not be deprecated), but I'll try again, with the following list of statements (used as abbreviations in the table below):
1. `Series.set_index` should exist, and accept array-likes
1. `DataFrame.set_index` should accept array-likes
1. existing ambiguity of `DF.set_index(list_of_scalars)` must be solved (orthogonal to #22225)
    1. issue warning in case of ambiguity
1. `DataFrame.set_index` should allow a mix of array-likes and column keys
1. `.set_axis` could be deprecated in favour of `set_index` 
    1. `DataFrame.set_columns` should be introduced for `axis=0` case (see also some discussion in #14829)
1. `set_axis` could simply switch between `set_index` and `set_columns` depending on `axis`-kwarg

The following table is my best attempt at summarizing the existing positions. Apologies for any mistakes, and feel free to edit "your" line in this table or comment (`+ / - / ?` are clear, and I use `~` as indifference):

@... thinks that ... | 1 | 2 | 3 | 3.i | 4 | 5 | 5.i | 6
------------ | --- | --- | --- | --- | --- | --- | --- | ---
@h-vetinari  | + | + | ~ | + | ~/+ | ~ | + | +
@jorisvandenbossche | + | + | ~ | ? | ? | + | ? | ? 
@jreback | ~? | -? | + | + | ? | ~? | ? | ?
@TomAugspurger | + | + | ~ | ? | + | ? | ? | ?
@toobaz  | ~? | ? | + | ? | -? | - | ? | -
@WillAyd | ~? | ? | + | ? | ? | ? | ? | ?

**Comment by TomAugspurger** [comment]
Thanks for the summary @h-vetinari. I think you've accurately summarized my views on 1, 2, and 3. I don't have any thoughts on changing `set_axis` right now.


>> I think it's quite useful to pass a mix of labels and keys.

> I would never do this in my code

I probably wouldn't either :) But I can see it being useful.

**Comment by toobaz** [comment]
> I probably wouldn't either :) But I can see it being useful.

By the way: this is the only reason I see for not deprecating/discouraging passing values to ``DataFrame.set_index``... and it is a use case that ``Series.set_index`` would _not_ cover.

**Comment by h-vetinari** [comment]
> @jreback: this needs coupling with possibly deprecating set_axis as well because passing values is not documented in any way

It *is* documented (albeit scarcely) in `v0.23.4` resp. [master](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.set_index.html):
> **keys** : *column label or list of column labels / arrays*

This parameter bullet has no description either, so that last "/ arrays" has to be counted as documentation for sure, IMO.

---------------------

> @toobaz: By the way: this is the only reason I see for not deprecating/discouraging passing values to `DataFrame.set_index`... and it is a use case that `Series.set_index` would not cover.

Why not make `DataFrame.set_axis(..., axis=0)` equal in capabilities to `DataFrame.set_index`? The main part is `set`ting the `index`, and yes, for `DataFrames` one can reasonably use existing columns instead of arrays. But how would this be contradicting the fundamental purpose of `set_axis`? I mean, a frame has two axes, right...? Assuming someone thinks "I want to `set` this `axis`", why not let them set it with the same capabilities as `set_index`?

To this end, I don't agree with your point further up the thread that having `set_axis(..., axis=0) == set_index` [and `set_axis(..., axis=1) == set_columns`] would violate the zen of python. Each of those methods has a clear and distinct purpose, but `set_index` is more specialised than `set_axis` (which is why one could easily dispatch from the latter to the former).

**Comment by toobaz** [comment]
> Why not make `DataFrame.set_axis(..., axis=0)` equal in capabilities to `DataFrame.set_index`?

OK, I think we're (I'm, at least) getting lost... so forget for a moment the current API, assume we start from scratch. We want the following:

1. set ``DataFrame.index`` from single column key
2. set ``DataFrame.index`` from sequence of values
3. set ``DataFrame.index`` from sequence of column keys
4. set ``DataFrame.index`` from sequence of sequences of values
5. mix 3. and 4.

All of these are good. Problem: a ``list`` of scalars is both a valid sequence of values (2.) and a potentially valid sequence of column keys (3.). There are two ways around the ambiguity: either (my preferred choice) we split functionalities 1. and 3. vs. 2. and 4. in two different methods, discarding 5., or we make an arbitrary choice and document it.

Now, back to the current API. We have ``set_index`` that does all of this, we have the ambiguity between 3. and 3., resolved in favor of 3.I don't like this, but other devs suggest it's not so problematic, that 5. is worth keeping, and certainly we can at least document everything better.

Then we have a method which does cleanly precisely what it is expected to: ``set_axis`` (it does 2. and 4.). Can we please at least agree on not making it ambiguous too, or deprecating it? Then we can continue the discussion on ``set_index`` ;-)

**Comment by h-vetinari** [comment]
> @toobaz: Now, back to the current API. We have `set_index` that does all of this, we have the ambiguity between 2. and 3., resolved in favor of 3. I don't like this, but other devs suggest it's not so problematic, that 5. is worth keeping, and certainly we can at least document everything better.

I think it's not *so* problematic if documented well. As you mentioned further up, `groupby` does similar gymnastics. I forgot to add that option to the table above.

> Then we have a method which does cleanly precisely what it is expected to: `set_axis` (it does 2. and 4.). Can we please at least agree on not making it ambiguous too, or deprecating it? Then we can continue the discussion on `set_index` ;-)

It's not central to the discussion here, but ideally, I'd like to solve the ambiguity *and* make `set_axis` have the same capabilities. Because setting an index IS setting an axis (and if you want to distinguish 1./3. from 2./4. by method, then it should be with a name that supports this distinction). Actually, my ordered preference for `set_axis` would be:
dispatch to `set_index` > deprecate `set_axis` entirely > keep as-is

----------------------

I do think it's worth considering to keep 5, but I can't see a good way to get rid of the ambiguity while keeping it. If we're **trying to get rid of the ambiguity first**, then I think a good solution could be to add an `arrays`-kwarg. The name "`keys`" is not an accurate description of what's happening anyway:

Docstring could look like:
```
def set_index(keys=None, arrays=None, drop=True, append=False, inplace=False,
              verify_integrity=False):
    """
    Set the index (row labels) using one or more given labels or arrays.

    keys : column label or list of column labels
    arrays : array-like or list of array-likes
    drop : ...
    [...]
    """
    if not (keys is None ^ arrays is None):
        raise ValueError('must pass exactly one of `keys`/`arrays`')
    # deprecate passing arrays as keys
```

**Comment by jreback** [comment]
When I evaluate an API change I try to achieve multiple objectives:

1) simple mental model
2) consistency

The most prevalent usecase ``DataFrame.set_index`` is one to set ``keys`` (from columns). This is a simple model and works well. Array values are *barely* documented and have a very limited set of test case, thus they are a rarely used case.

Adding ``Series.set_index`` would shift the primary use case to be setting with ``values``. This IMHO violates consistency (2) because now if you are working with a ``DataFrame`` you have different primary use-cases. This violates the mental model (1) as we now have 2 methods that do *exactly* the same thing for ``Series``, meaning ``.set_axis``, it also violates (2) as you now *still* have to know whether you are working with a ``Series`` or a ``DataFrame``, which I gather is the impetus for having ``Series.set_index`` in the first place.

I don't see a good way of reconciling this and thus don't see a good case of adding ``Series.set_index``. (I had suggested a possible way forward of *deprecating* ``Series.set_axis`` to make room for ``Series.set_index``, which would fix my objection, but still leave us with inconsistencies between ``DataFrame`` and ``Series`` for use case. Adding keyword args to ``.set_index`` is not a solution as this makes things more complicated.

As @toobaz notes above, if we were to start fresh we might do it differently, but back-compat here is paramount as its not very easy to change existing behavior. 

Finally I would consider removing support for allowing arrays in ``DataFrame.set_index``. For the small use cases where this is actually useful there are quite a lot of *good* solutions that did not exist when pandas first existed, e.g.

``DataFrame.assign(new_col=.....).set_index(['A', 'new_col'])`` is quite idiomatic.

**Comment by TomAugspurger** [comment]
How do we know that setting arrays with DataFrame.set_index isn't common?

Would you say that, from a consistency point of view, accepting an array
for Series.set_index is *more* consistent
than not?

**Comment by h-vetinari** [comment]
@jreback 
Thanks for taking the time for a comprehensive answer.

> The most prevalent usecase DataFrame.set_index is one to set keys (from columns). This is a simple model and works well. Array values are *barely* documented [...]

I agree with @TomAugspurger that we can't know the prevalence a-priori. `df.set_index(df2.index)` is simply too intuitive, and *just works* (assuming the same length), without having a look in the docs at all. Same for `df.set_index([df2.index, df3.some_column])`.

> [...] and have a very limited set of test case, thus they are a rarely used case.

It is extensively tested, already before #22236 (but even more extensively since that PR) and #22486. In particular, all list-likes (except sets) are tested, and all combinations between list-likes, also crossed with column labels (i.e. case 5. of @toobaz' list above).

> As @toobaz notes above, if we were to start fresh we might do it differently, but back-compat here is paramount as its not very easy to change existing behavior.

Of course I'm thinking of back-compat as well. How about the suggestion above to turn `keys` and `arrays` into kwargs? This would have to deprecate the mixed case, but would remove the ambiguity at least.

> Finally I would consider removing support for allowing arrays in `DataFrame.set_index`. For the small use cases where this is actually useful there are quite a lot of good solutions that did not exist when pandas first existed, e.g.
`DataFrame.assign(new_col=.....).set_index(['A', 'new_col'])` is quite idiomatic.

Won't that cause extra copying in `assign`? In any case, deprecating *mixes* of keys and arrays might be a good idea, but I strongly believe that arrays should not be deprecated (see above).

**Comment by jreback** [comment]
> How do we know that setting arrays with DataFrame.set_index isn't common?

easy enough, its barely documents, and I have *never* seen a bug report about this.

> Would you say that, from a consistency point of view, accepting an array
for Series.set_index is *more* consistent
than not?

no, this is very confusing because it is *different* than the primary usecase for ``Dataframe.set_index``.

@h-vetinari 

> I agree with @TomAugspurger that we can't know the prevalence a-priori. df.set_index(df2.index) is simply too intuitive, and just works (assuming the same length), without having a look in the docs at all. Same for df.set_index([df2.index, df3.some_column]).

Show me where this is in the documentation

> Of course I'm thinking of back-compat as well. How about the suggestion above to turn keys and arrays into kwargs? This would have to deprecate the mixed case, but would remove the ambiguity at least.

no this is stricty worse and more confusing.

> Won't that cause extra copying in assign? In any case, deprecating mixes of keys and arrays might be a good idea, but I strongly believe that arrays should not be deprecated (see above).

I am not concerned about performance; being consistent is paramount. I am not strongly for deprecating, though I think it would be better, because as I say we *already* have a way to do this, ``.set_axis``

**Comment by h-vetinari** [comment]
> @jreback: Show me where this is in the documentation

[0.12](https://pandas.pydata.org/pandas-docs/version/0.12/generated/pandas.DataFrame.set_index.html) - [0.20.3](https://pandas.pydata.org/pandas-docs/version/0.20.3/generated/pandas.DataFrame.set_index.html) (very likely even further back; 0.12 is the oldest one online):
```python
>>> indexed_df = df.set_index(['A', 'B'])
>>> indexed_df2 = df.set_index(['A', [0, 1, 2, 0, 1, 2]])
>>> indexed_df3 = df.set_index([[0, 1, 2, 0, 1, 2]])
```

[0.21.0](https://pandas.pydata.org/pandas-docs/version/0.21.0/generated/pandas.DataFrame.set_index.html) - [master](https://pandas-docs.github.io/pandas-docs-travis/api/generated/pandas.DataFrame.set_index.html) (last example):
> Create a multi-index using a set of values and a column:
```python
>>> df.set_index([[1, 2, 3, 4], 'year'])
         month  sale
   year
1  2012  1      55
2  2014  4      40
3  2013  7      84
4  2014  10     31
```

This makes me think that even @toobaz' case 5 might be too entrenched to deprecate, but what could also be done is just **deprecating `list-likes`** that are not `array-like` (like for `str.cat` in #22264). I think this is my favourite solution so far, because it solves the ambiguity while maintaining essentially all functionality (i.e. passing keys / ndarray / Series / Index, or a list-like combination thereof).

Further options would be to just clearly document the current "list-gets-interpreted-as-keys", or add a `groupby`-like fall-back that a list without any keys will be used as an array. In all three cases, there is no conflict with `Series.set_index` supporting arrays.

> @jreback: [...] I have never seen a bug report about this.

That could just as well be because it always either works or gives a clear error that the length does not match. (also see above: implemented since forever)

```python
>>> pd.DataFrame(np.eye(2)).set_index(pd.Series([1, 2, 3]))
[...]
ValueError: Length mismatch: Expected axis has 2 elements, new values have 3 elements
```

**Comment by h-vetinari** [comment]
Let me extract that last idea a bit, lest it get lost in the larger post above:

How about deprecating `list`-likes that are not `array`-like? This would be:
* the least invasive change
* resolve the ambiguity of `list_of_scalar`
* retain the existing and long-documented capability to:
    * use arrays
    * mix column keys with arrays

@jreback @jorisvandenbossche @TomAugspurger @toobaz @WillAyd

**Comment by h-vetinari** [comment]
Almost complete implementation of that last suggestion is in #24697, to better be able to discuss this.

**Comment by toobaz** [comment]
> How about **deprecating `list`-likes that are not `array`-like**? This would:

I already replied in #24697 ... but I'm afraid more in general your proposal doesn't solve the ambiguity nicely. If in "`list`-likes" you include lists, then

```
df.set_index([['a', 'b']])
```
now doesn't mean _anything_... which is (I guess) worse than it meaning "take these values". If in "`list`-likes" you _exclude_ lists (which, if I understand correctly, is what you do in #24697), then the ambiguity is unaffected.

**Comment by h-vetinari** [comment]
> @toobaz: If in "`list`-likes" you include lists, then `df.set_index([['a', 'b']])` now doesn't mean anything

That is indeed the idea (and how it is implemented in #24697). I'd say it's unclear what  `df.set_index([['a', 'b']])` is even supposed to mean (for people who don't know the nested-list trick), and the double braces are easy to overlook.

Lists, like tuples, are used for too many things IMO (causing ambiguities, "magic" methods, and code complexity). I'd also postulate that outside of toy examples, they're hardly used for this either. In most real-world cases, people already have Series/Index/arrays that they're juggling.

We've had a similar deprecation in #22264 for `str.cat`, because it will make the code much easier to maintain, and does not sacrifice any essential functionality (it's very easy to wrap a correctly-sized list into an array/Series etc.).

**Comment by h-vetinari** [comment]
After not getting a response in #24697, I'd like to ask again how to proceed here. Deprecating lists within lists would have IMO been the most elegant solution, and was the only complaint against adding `Series.set_index`. Now that the decision has been taken to live with this ambiguity, can we please reopen #22225?

**Comment by pilkibun** [comment]
I waded into this before learning how contentious it is. I would like to propose the following compromise:

1. Add `Series.set_index` with behavior matching current `DataFrame.set_index` with respect to arrays. List of lists not allowed.
2. Immediately deprecate this case in both DataFrame/Series, and issue a deprecation warning when Arrays are used, telling users to use `set_axis` instead.
3. keep `set_index` for column keys forever. keep `set_axis` for setting labels forever.
4. after a deprecation cycle, remove the functionality from `set_index`

The benefits are:
1. resolves the issue
2. reduces inconsistency between Frame/Series and offers a clear transition path to separating the two, while guiding users to the alternative.
3. after a deprecation cycle, leads to a state which satisfies everyone: no ambiguity, and both use made easy for users.

It is vitally important to have a method-chaining friendly way to set a series index. I'd like to help make that happen. note that `set_axis` is not very well known currently, and appears in the documentation only in the API reference section. If acceptable, I can fix that as well.

**Comment by h-vetinari** [comment]
> @pilkibun: I waded into this before learning how contentious it is.

Haha, me too. ;-)

Actually, a lot happened after #22225, namely the discussions in #24046 and #24697 are pertinent. The principal bone of contention (from my POV) was how to deal with the ambiguity of `list_of_scalars` being interpreted as an `array-like` vs. as a `list_of_column_keys`. Personally, I would have deprecated `list_of_scalars_as_an_array`, but that ultimately got rejected after non-public discussions of the core-devs.

In any case, the issue remains an sorely lacking feature IMO, as users are (and should be) discouraged to directly manipulate attributes (e.g. `s.index = <index>`), and `Series.set_index` is doubtlessly the most intuitive method for `set`ting the `index` (and also missing as an analogue for `DF.set_index`). I do not think that `set_axis` is an adequate replacement.

So far, all mentioned proposals boil down to the following:
1. do nothing
1. add `Series.set_index` as in #22225 but live with the ambiguity of `list_of_scalars` (resp. resolve that ambiguity towards `list_of_column_keys` consistently)
1. deprecate `list_of_scalars_as_an_array` as in #24697 and then add `Series.set_index`
1. deprecate using `set_index` with arrays, and point to `set_axis` instead [if I understand you correctly]

To me, only 2 & 3 are reasonable solutions vis-à-vis user-friendliness.

**Comment by toobaz** [comment]
I'm still in favor of 4, but I admit I got totally lost in the discussion(s)

**Comment by WillAyd** [comment]
Yes I think option 4 as well

**Comment by pilkibun** [comment]
What I'm suggesting is, for now, to ensure as much parity between DataFrame and Series `set_index`. If   a decision is later made to move some functionality elsewhere, it's no more difficult to deprecate both simultaneously. And If no agreement can be reached on that, it's acceptable to have the functionality in both.

Note that `set_index(array, append=true)` and `set_index([array, array])` both create MultiIndexes, which is useful, and not currently possible with `set_axis`. We could add that later of course as prep for splitting it off. But even if we do, there will be a deprecation cycle, and in the meantime it's not much worse than having arrays this in `DataFrame.set_index`.

I have put together #27504 that closely follows `DataFrame.set_index`  and I'd like your support for it. +1/-1 for this suggested plan of action, please?

**Comment by toobaz** [comment]
>  If a decision is later made to move some functionality elsewhere, it's no more difficult to deprecate both simultaneously

@pilkibun feel free to tell me if I misunderstood, but we certainly won't allow a feature which we plan to (probably) later deprecate

**Comment by h-vetinari** [comment]
> @toobaz: I'm still in favor of 4, but I admit I got totally lost in the discussion(s)

> @WillAyd: Yes I think option 4 as well

I think whichever future pandas version were to remove arrays for `set_index` would receive massive pushback - it is one of the most intuitive parts of the API, and just works.

Also, index *is* an axis, so not least from the point of consistency, I'd say that `set_axis` should actually have the same capabilities as `set_index` (with the only difference being an `axis`-switch). It would be confusing (i.e. opposite of intuitive) to remember which of those extremely similar methods does what, if option 4 were pursued.

**Comment by toobaz** [comment]
> I think whichever future pandas version were to remove arrays for `set_index` would receive massive pushback - it is one of the most intuitive parts of the API, and just works.

It's trivial to replace with ``set_axis``, which also just works. The difference (again, if I remember correctly) is just that you cannot mix labels and arrays, but I really don't expect this to be a frequent use case.

In any case, I've never seen ``set_axis`` used with arrays in real code. I won't say it never happens, but I definitely think it is a marginal use case.

> Also, index _is_ an axis, so not least from the point of consistency, I'd say that `set_axis` should actually have the same capabilities as `set_index` (with the only difference being an axis-switch).

"consistency" is not to have duplicated methods with different names. The day that ``set_axis`` was really just a subset of ``set_index``, we would probably just deprecate it entirely ;-)

What is definitely not consistent is that we have ``set_index`` but not ``set_columns``. Not a huge problem, but another reason why we maybe don't want to sponsor ``set_index`` over ``set_axis``.

**Comment by h-vetinari** [comment]
@toobaz, you're still mixing up a few things (or maybe I need to write more clearly...? or you mistyped...?)

> In any case, I've never seen `set_axis` used with arrays in real code. I won't say it never happens, but I definitely think it is a marginal use case.

The point is using arrays with `set_INDEX`, and keeping that use, which is certainly not marginal. Not least because it is, by a country mile, the first thing people turn to to set the index (which is very reasonable with an array of the same lengths as the Series/DF).

> It's trivial to replace with `set_axis`, which also just works. The difference (again, if I remember correctly) is just that you cannot mix labels and arrays, but I really don't expect this to be a frequent use case.

Trivial, but harmful, because it would be removing the most obvious method for the job (sacrificing this usability should require something that's demonstrably *better*- and I'd challenge that option 4 satisfies that).

Mixing column labels and arrays might actually be a marginal use case, but it's been documented since before v0.12, so I'll reserve judgement. It would be much less impact to deprecate this aspect only, but that does not solve the ambiguity of `list_of_scalar`.

> "consistency" is not to have duplicated methods with different names. The day that `set_axis` was really just a subset of `set_index`, we would probably just deprecate it entirely ;-)

It would be a *super*set (as in, encompassing both `set_index` and a putative `set_columns`, and could just be a switch between those). Although `set_axis` would be the first of those three that I'd get rid of, if necessary, I think that setting an index/column/axis are all eminently obvious and useful operations, that deserve their own method. And since they perform the exact same task, they should also have the same interface (as far as reasonably possible; i.e. no column keys for Series).

> What is definitely not consistent is that we have `set_index` but not `set_columns`.

I'd argue (both above and in the OP), that this method should be added.

--------------------------

TLDR: An argument for option 4 would have to demonstrate a much greater user benefit for removing such fundamental functionality as using arrays in `set_index`. So far, the only reason given was the (IMO much weaker argument) that `list_of_scalars` is ambiguous, which could be solved with much less impact by deprecating `list_of_scalars_as_array` (i.e. option 3).

**Comment by toobaz** [comment]
> or you mistyped...?)

Yeah, I meant ``set_index``. You say ``set_index`` is the first thing people turn to, and I totally agree. Then you say that doing it with arrays is "reasonable"... which is different ;-), and there's already a method for that.

> It would be a _super_set

No,"deprecate it" = deprecate ``set_index``. (By the way, it was purely hypothetical reasoning)

For the rest, you make some confusion between removing functionality (the only functionality removed in my proposal would be the mixed case, which you admit is not so important) and just making the API more tidy, asking some users to replace a method name.

I'm all against 3 because I'm all against special cases (and I think I didn't even take part in the discussion mentioned in https://github.com/pandas-dev/pandas/pull/24697 !)

**Comment by pilkibun** [comment]
FWIW, yes, I'm in favor of making `Series.set_index` behave like `DataFrame.set_index`, it's useful, and it's where users look for this functionality. I also wouldn't lose sleep if it were deprecated as long as set_axis gained abilities to compensate.

I see absolutely no harm in adding `Series.set_index` under any future scenario. Certainly if we never deprecate, and there's not real problem to deprecate both when that happens.

The really disappointing aspect of this whole thing is that there's been no progress to do either for over a year (two?). When I submitted a PR to document, it was politely declined because, or so it seems to me,  the decision on this has been permanently deferred. So I'm ok with *either* solution, but *neither* seems likely to happen as things stand.

Anyway, I should apologize for restarting this discussion. It was incredibly naive of me to think I could make a difference where so many have tried before and now I'm letting this go.

**Comment by h-vetinari** [comment]
> @toobaz: For the rest, you make some confusion between removing functionality [...] and just making the API more tidy

Well, `df.set_index` *currently* supports arrays, and (IIUC) you want to remove that functionality. 

> @toobaz: Then you say that doing it with arrays is "reasonable"... which is different ;-), and there's already a method for that.

I disagree that it's unreasonable/different, and also that `set_axis` can serve as a direct replacement; because it's not obvious except to people who know the arbitrary idiosyncrasy that for using arrays, they have to switch to `set_axis`.

> toobaz: I'm all against 3 because I'm all against special cases (and I think I didn't even take part in the discussion mentioned in #24697 !)

The "special-case" objection is fair enough, but it has to be weighed against the impact of other alternatives, namely ambiguity of `list_of_scalar`, or removing core functionality (i.e. using arrays in `set_index`). To me, both of the latter are worse resp. much worse than having a special case for the list, which has to be interpreted in too many places (leading to _other_ special cases). This is very similar to how tuples are also overused (and about tuples you agree, AFAIK... #24702).

**Comment by h-vetinari** [comment]
> @pilkibun: Anyway, I should apologize for restarting this discussion. It was incredibly naive of me to think I could make a difference where so many have tried before and now I'm letting this go.

IMO you shouldn't apologize!
1. you couldn't have known [realistically speaking]
1. it's important to draw attention to these deficiencies, no matter how entrenched. Just because a certain corner of the API is difficult/delicate shouldn't mean it cannot ever be rectified/improved.

So from my side: thanks for having picked this up again (however briefly). ;-)

**Comment by toobaz** [comment]
About terminology: let's not talk about "removing core functionality" when we mean reshaping the API by removing duplicates. The only functionality removal we are talking about here is the mixed case. Only by avoiding provocations we can try to reduce the sheer amount of noise in this issue.

> To me, both of the latter are worse resp. much worse than having a special case for the list, which has to be interpreted in too many places (leading to _other_ special cases). This is very similar to how tuples are also overused (and about tuples you agree, AFAIK... #24702).

I think you have a point if you are saying (not sure I understood correctly) that we tend to represent ``MultiIndex`` keys with tuples, and that the content of nested lists are ``MultiIndex`` keys. But I think there is a stronger point: where we accept arrays, we also accept lists; and if we accept 2D arrays, we will accept lists of lists. In other words, before data becomes the content of a ``MultiIndex``, it is "just data".

**Comment by h-vetinari** [comment]
> @toobaz: Only by avoiding provocations we can try to reduce the sheer amount of noise in this issue.

There's no provocation; I will however correct it if my argument is misrepresented (unintentionally).

> About terminology: let's not talk about "removing core functionality" when we mean reshaping the API by removing duplicates. The only functionality removal we are talking about here is the mixed case. 

I'm very sorry, but it *is* removing core functionality - namely "arrays in `set_index`". It was an enhancement when it was [added](https://github.com/pandas-dev/pandas/commit/553bacc7abe341991e4d043a5c17521902c7f127), and 7 years later, it is justifiable to call it core functionality (esp. considering how pandas is *all* about indexed data).

Talking about `set_index` and `set_axis` as "duplicates" is flawed IMO, because they deal with the exact same objects ("axes") - they are much less duplicates, rather than two sides of the same coin. If you want to deduplicate *on the implementation side*, you can make `set_axis([...], axis=0)` use the implementation of `set_index([...])`.

TLDR: Having different behaviour for `set_index` and `set_axis` is IMO bad API-design.

> I think you have a point if you are saying [...]

My point was mostly different: lists - like tuples - are used in too many different _roles_, forcing many code paths for all the different scenarios in many places. Just as I support interpreting tuples *always* as scalars (and not, e.g. as array-likes; cf. #24702), I would argue that lists should preferably have the role of a sequence of objects (as in: "these columns are supposed to form a MultiIndex"), and also give up the array-like interpretation (which can be trivially replaced by Series/Index/np.ndarray/etc.).

**Comment by h-vetinari** [comment]
I forgot to add a fifth option to the recap above:
Turning the signature from
```
def set_index(keys, drop=True, append=False, inplace=False, verify_integrity=False):
```
to
```
def set_index(keys=None, arrays=None, drop=True, append=False, inplace=False,
              verify_integrity=False):
```
as [suggested](https://github.com/pandas-dev/pandas/issues/24046#issuecomment-452613625) further up the thread. This would also sacrifice the mixed case, but make it very explicit what's being demanded (which would also resolve the ambiguity of `list_of_scalar`).

PS. @pilkibun, adding materially new content to a comment through an edit makes it hard to see resp. respond to.

**Comment by pilkibun** [comment]
Sorry, I removed the new content and also most of the old. 

Without arguing for one position over another, here's the state of things as seen from the user's point of view who looks to the documentation for guidance:
```
# grep -R '\.index' doc/source
whatsnew/v0.14.0.rst
158:    df_multi.index = tuple_ind
164:    df_multi.index = mi

whatsnew/v0.16.0.rst
98:   s.index = pd.MultiIndex.from_tuples([(1, 2, 'a', 0),

user_guide/indexing.rst
1706:   data.index = index

getting_started/10min.rst
653:   ts.index = (prng.asfreq('M', 'e') + 1).asfreq('H', 's') + 9

user_guide/io.rst
1893:   dfj2.index = pd.date_range('20130101', periods=5)
2952:   df.index = df.index.set_names(['lvl1', 'lvl2'])

getting_started/basics.rst
233:   dfmi.index = pd.MultiIndex.from_tuples([(1, 'a'), (1, 'b'),

user_guide/timeseries.rst
2096:   ts.index = (prng.asfreq('M', 'e') + 1).asfreq('H', 's') + 9

user_guide/sparse.rst
306:   s.index = pd.MultiIndex.from_tuples([(1, 2, 'a', 0),
```

At the same time, all example uses of `set_index` are for the column case and there are 0 usage examples of `set_axis` which appears only in the auto-generated reference. 

I think it's fair to say that the docs currently advocate for always setting the index directly and gives  no indication that setting a series index in a method-chain is supported. I care about this more than I care about which competing alternative wins out. 

My personal preference is still to keep `set_index` as it is and add its equivalent to `Series`. And in any case to update the documentation accordingly.

**Comment by toobaz** [comment]
> I'm very sorry, but it _is_ removing core functionality

You're not trying to understand.

**Comment by toobaz** [comment]
> give up the array-like interpretation

I'm pretty sure this will never pass.

**Comment by h-vetinari** [comment]
> > @h-vetinari: I'm very sorry, but it *is* removing core functionality

> @toobaz: You're not trying to understand.

I understand that, to you, the functionality is not removed because it lives on in `set_axis`. We just disagree about what constitutes good API design.

> > @h-vetinari: give up the array-like interpretation

> @toobaz: I'm pretty sure this will never pass.

It doesn't have to be everywhere or at the same time, but removing such overloaded interpretations would greatly simplify the API surface as well as code maintainability. I'm saying it could be a goal to strive for, like striving to have tuples always be MI-keys (and was an argument why "special-casing" in option 3 can be desirable).

> @pilkibun: [...] and there are 0 usage examples of `set_axis` which appears only in the auto-generated reference.

@toobaz, this is part of the reason why I consider shifting the array-capability from `set_index` to `set_axis` as a removal of functionality - it is not nearly as intuitive or wide-spread as `set_index`.

**Comment by toobaz** [comment]
> We just disagree about what constitutes good API design.

We mostly disagree on what constitutes productive interaction. And from my part, this is the last "metacomment" in this discussion.

> removing such overloaded interpretations would greatly simplify the API surface as well as code maintainability

I totally agree it would simplify the code, but we don't want our users to pay the price. And it would be particularly ironic to start doing so because we _want_ to overload the interpretation of the argument to ``set_index``.

So among the options you [listed](https://github.com/pandas-dev/pandas/issues/24046#issuecomment-513914495), only 4 and 5 are feasible. To be honest I don't like 5, but this might be subjective. Do you agree that a deprecation message which tells our users "please replace set_index with set_axis" will serve our users at least as well as a deprecation message which tells our users "please replace set_index with set_index(arrays=.)"?

**Comment by h-vetinari** [comment]
> I totally agree it would simplify the code, but we don't want our users to pay the price. And it would be particularly ironic to start doing so because we want to overload the interpretation of the argument to set_index.

I agree that users should not pay the price. I also don't want the overloaded interpretation in `set_index`, but would tackle it differently.

> Do you agree that a deprecation message which tells our users "please replace set_index with set_axis" will serve our users at least as well as a deprecation message which tells our users "please replace set_index with set_index(arrays=.)"?

As a deprecation, yes. After the removal, not so much (because `set_index` will still be the more obvious choice, also for arrays).

> So among the options you listed, only 4 and 5 are feasible.

I think 3 and 5 are feasible. The overlap is 5, and I'd be happy to support that.

> We mostly disagree on what constitutes productive interaction.

Indeed, and we both have our failings there. Thanks for taking the time/energy to stick with it.

**Comment by toobaz** [comment]
> because `set_index` will still be the more obvious choice, also for arrays

Why "obvious"? For the name "``set_index``" vs. "``set_axis``"? Or because you like everything in the same method?
(either way, I think it's just a matter of having clear documentation)

> I think 3 and 5 are feasible.

3 was so far ruled out by _both_ me and @jreback - not very useful to cite it if you don't have new arguments. I guess we never discussed much 5, so I'll be happy to mention it in  the next live chat. I still think we will end up choosing 4.

**Comment by h-vetinari** [comment]
> Why "obvious"? For the name "`set_index`" vs. "`set_axis`"?

That plays a large role, yes, because we're `set`-ting the `index`-attribute, and because having an intuitive API is crucial. Documentation does not help with intuitiveness.

> 3 was so far ruled out by *both* me and @jreback - not very useful to cite it if you don't have new arguments. 

I have enough arguments**, but so far you have hardly responded to them, except by appeal to authority. 4 or 5 is your opinion, 3 or 5 is mine. Of course, you're in a much better position to enforce your ideas, but that does not improve the strength of your argument.

Still, we have already found some sort of minimal common ground with 5. If your preference for 4 over 5 is not too large to be overcome at all, I'd be happy to submit a PR that implements 5.

** I realised I haven't even fully articulated one of the most important arguments against 4: saying that arrays can only be used in `set_axis` **does not solve the lack of a `set_index`-method for Series**! The rest of my main arguments is recapped below the fold:
<details>

* that "arrays-in-`set_index`" was a specific [enhancement](https://github.com/pandas-dev/pandas/commit/553bacc7abe341991e4d043a5c17521902c7f127) and that this is core functionality not least due to its intuitiveness and centrality of indexes to pandas
* that axis/index are two sides of the same coin (and not two duplicate coins), and that having different behaviours between `set_index` and `set_axis` is further detriment to intuitiveness
* that deprecating `list-as-array` (vs. `list-as-collection`) is a goal to strive for, much more than a special case

</details>

**Comment by toobaz** [comment]
> appeal to authority

This is a way to waste the time of both of us. So far, the only effect in this discussion of me being a core dev is that I am patiently replying to your rants, instead than doing more productive things.

But since you later repeated your arguments in a tidy way, I will for the last time to assume you're trying to be constructive, and repeat my objections to your arguments. But neither your arguments nor my replies are new - and my replies are not just mine, other devs contributed to this and related discussions. If you want to just discuss the points below over and over, let's do this in a live chat so at least we wast less time.

> that "arrays-in-`set_index`" was a specific [enhancement](https://github.com/pandas-dev/pandas/commit/553bacc7abe341991e4d043a5c17521902c7f127)

The fact that an ability is added later on does not mean it cannot be deprecated (actually the opposite).

> and that this is core functionality not least due to its intuitiveness and centrality of indexes to pandas

"setting index with arrays" is a functionality (not so sure it is even a "core" one - setting from columns is much more frequent in the code I see), having it in ``set_index`` is just a matter of organizing the API. ``axis`` is not a strange term in pandas, it is one of the core concepts, so while ``index`` might be even more immediate, we are definitely not hiding the functionality by keeping it in ``set_axis``.

> that axis/index are two sides of the same coin (and not two duplicate coins), and that having different behaviours between `set_index` and `set_axis` is further detriment to intuitiveness

Your argument would suggest we want to make them almost-duplicates (apart from the ``axis=`` argument), and this is precisely something any good API should avoid. Vice-versa, separating the functionalities allows for less ambiguity (no, "ambiguity" is not an argument against option 5, but having a single method do widely different things based on arguments is not a favor to our users).

> that deprecating `list-as-array` (vs. `list-as-collection`) is a goal to strive for, much more than a special case

As already clearly stated: you are right this would simplify our life, but we don't want our users to pay for. We always allowed users to skip the step of explicitly creating a vectorized object to feed our vectorized objects, and I see no reason why we should change our mind now. As for terminology, 2D arrays _are_ collections of 1D arrays, which are collections of elements, so lists of lists make perfect sense.

> saying that arrays can only be used in `set_axis` **does not solve the lack of a `set_index`-method for Series**

There is nothing to "solve" if the ``set_index`` does something which is completely useless for ``Series`` - we have ``set_axis`` that does the useful part. And we have discussed this some time ago. This said, if we do find that some sort of ``Series.set_index`` is useful for compatibility (like the ``MultiIndex`` methods which were backported to flat ``Index``es as idempotent), I will have no general objections. But it will be a result of, not an argument for, our decision about ``DataFrame.set_index``, and it is OT here.

**Comment by h-vetinari** [comment]
> So far, the only effect in this discussion of me being a core dev is that I am patiently replying to your rants, instead than doing more productive things. [...] I will for the last time to assume you're trying to be constructive, [...]

I appreciate the time you took, and thanks for that. Indeed I was trying to have a productive discussion, and what seemed like rants to you was my response to the perception that you dismissed my points out of hand (until your last reply).

> Your argument would suggest we want to make them almost-duplicates (apart from the `axis=` argument), and this is precisely something any good API should avoid.

The point is that the concepts `axis` and `index` are "almost-duplicates" already, and so their methods should reflect that (for good design, at least). Thinking of the _methods_ as duplicate is having things backwards, because the concept comes before the method.

> We always allowed users to skip the step of explicitly creating a vectorized object to feed our vectorized objects, and I see no reason why we should change our mind now.

Except maybe if it is the cause of the hotly-contested ambiguity of `list_of_scalar` that stalled this whole discussion since about a year. It's clear that there are no great choices here, but leaving users to decipher the difference between `set_index([0, 1, 2])` and `set_index([[0, 1, 2]])` is not amazing either.

> This said, if we do find that some sort of `Series.set_index` is useful for compatibility (like the `MultiIndex` methods which were backported to flat `Index`es as idempotent), I will have no general objections. But it will be a result of, not an argument for, our decision about `DataFrame.set_index`, and it is OT here.

I would hope this were OT, but this issue was a direct consequence of #22225 (simply adding `Series.set_index`) being rejected on the grounds of the _existing_ capabilities of **`DataFrame`**`.set_index`.

________________________________

Anyway, thanks for taking the time to respond. I'll note that you didn't react to my attempt at extending an olive branch though:

> @h-vetinari: Still, we have already found some sort of minimal common ground with 5. If your preference for 4 over 5 is not too large to be overcome at all, I'd be happy to submit a PR that implements 5.

**Comment by toobaz** [comment]
> The point is that the concepts `axis` and `index` are "almost-duplicates" already, and so their methods should reflect that

In API design, "adherence to language" is only one of the many arguments - and an argument which is not new in this discussion. And for sure the fact that we have (almost-)duplicates in language doesn't mean we want (almost-)duplicates in the API.


> the hotly-contested ambiguity of `list_of_scalar` that stalled this whole discussion since about a year

You seem to think that this problem is important because it stalled for a year - maybe it stalled for a year because there were more important things ;-)

> It's clear that there are no great choices here, but leaving users to decipher the difference between `set_index([0, 1, 2])` and `set_index([[0, 1, 2]])` is not amazing either.

Flat lists and nested lists _should_ be clearly different objects to our users.

> I'll note that you didn't react to my attempt at extending an olive branch though

I did. I [said I don't like 5](https://github.com/pandas-dev/pandas/issues/24046#issuecomment-514939649), I [said why](https://github.com/pandas-dev/pandas/issues/24046#issuecomment-515360096), but I also [said I'm happy to discuss at our next live chat](https://github.com/pandas-dev/pandas/issues/24046#issuecomment-514984921).

Now, unless there are _new_ arguments, I suggest we wait for that.

**Comment by h-vetinari** [comment]
> And for sure the fact that we have (almost-)duplicates in language doesn't mean we want (almost-)duplicates in the API.

Then the consequence should be deprecating `set_axis`, but not having very different methods for almost duplicate concepts.

> [...] but I also said I'm happy to discuss at our next live chat.

Is there a date/time already?

**Comment by toobaz** [comment]
> Then the consequence should be deprecating `set_axis`, but not having very different methods for almost duplicate concepts.

Setting the index from columns and from data are two operations (I guess this is what you mean by "concepts") which are different enough (we have seen) as to raise the need to avoid ambiguity. It can be done through different functions, or different args, or by defining special cases users should be aware of, but we definitely agree they are distinct. Which of these solutions we pick is mostly not a matter of linguistics.

> Is there a date/time already?

No

**Comment by toobaz** [comment]
> No

(but just write me privately if you want to set up a chat with me)

**Comment by wesm** [comment]
I'd like to point out that the tone of this thread makes me a bit uncomfortable. As a reminder, this project has a code of conduct

https://github.com/pandas-dev/pandas-governance/blob/master/code-of-conduct.md

In such discussions, I think we (both maintainers and contributors) need to stick to facts and technical arguments and leave feelings and editorial comments out of the process. There is a risk in technical arguments to stoop to emotive conjugation (https://en.wikipedia.org/wiki/Emotive_conjugation) in describing others' actions.

In general my understanding is that this project operates on the basis of consensus-based decision making -- when there is no consensus about a change, the default option is probably to do nothing. In theory as the BDFL I can help settle disagreements, but I would prefer not to except in truly exceptional circumstances. 

I question whether GitHub issues was the appropriate venue for this discussion compared with some form of RFC / design document.

**Comment by h-vetinari** [comment]
@wesm 
Thanks for taking the time to respond here, although I regret that the reason was due to discomfort.

I have striven to avoid any emotionally charged words, but don't claim that I always succeeded. I believe all participants truly want the best for the combination of user- & maintainer-base, but such impassioned arguments take a lot of time and energy (which, I presume is the reason why many participants have not joined the discussion anymore).

I do object to the way some things were handled in this whole episode, but will not dwell on that. My main reasons for not resigning from this discussion are that I don't want the case dismissed without a fair hearing/counter-argument (even if I'm not core-dev), and that I feel the picture is [much less one-sided](https://github.com/pandas-dev/pandas/issues/24046#issuecomment-452050651) even on the dev-side, as the impression that last few comments might give.

> I question whether GitHub issues was the appropriate venue for this discussion compared with some form of RFC / design document.

I'd be happy to participate in another format, but didn't know a better way that through an issue here.

PS. Thanks for the link about emotive conjugation. "How would I describe myself in their shoes" will be an excellent self-check before speaking/posting.

**Comment by toobaz** [comment]
In the last dev chat, which was held just few minutes ago, this issue was discussed and there was clear consensus for option 4, that is, "deprecate using set_index with arrays, and point to set_axis instead".

(Related to the "tone" of the discussion, I definitely try to keep emotionally charged language away from my comments, but whenever should I fail to do so, I welcome being explicitly reprimanded - ideally in private. I'm not a native speaker, and in any case I definitely won't be offended by any such rebuke. On the other hand, language aside, I think that when a discussion takes more of our time and energy than it is worth, there is nothing wrong in stating it, even if it is not a "fact or technical argument" on pandas itself.)

**Comment by jorisvandenbossche** [comment]
@toobaz can you give some reasoning why this is the clear preference? 
I would need to go through this long thread the understand it more, but maybe some arguments were summarized on the call?

**Comment by h-vetinari** [comment]
@toobaz 
Thanks for seeing this through. Although I would have liked to participate in the dev chat about this, and although I find the decision suboptimal, I guess any decision is better than no decision at this point. Time permitting, I'll look into a PR that deprecates arrays from set_index and outputs a nice warning to use set_axis.

@jorisvandenbossche 
The lack of documentation and transparency (note: not an accusation, that's just the way it has been so far) in such cases is why I'm thinking about a pandas version of PEPs/NEPs (#28568). I have been swamped recently and couldn't respond on that issue, but I will pick it up again.

**Comment by toobaz** [comment]
> @toobaz can you give some reasoning why this is the clear preference?

To be honest there was more a recall of the reasoning already exposed here than any new argument. Some devs (e.g., @WillAyd ) were already clearly in favour of option 4.

The only new thing was a proposal (by @TomAugspurger if I recall correctly) to deprecate ``set_index`` entirely, replacing with two methods ``set_index_keys`` and ``set_index_values``, a clean, but more disruptive, solution. But in the end, consensus on 4 was reached pretty quickly.

@h-vetinari you probably already know, but just in case: the call was publicly announced on the [pydata] mailing list. In any case, again, if there had been new arguments I would have written them here. It is hard to see the 76 comments here + other in related issues a "lack of documentation and transparency".

>  I'm thinking about a pandas version of PEPs/NEPs (#28568).

Wlll reply there.

**Comment by h-vetinari** [comment]
@toobaz 
I didn't know the announcement, thanks for the info.

> @toobaz: It is hard to see the 76 comments here + other in related issues a "lack of documentation and transparency".

I do not consider dispersed discussion in several threads and comments as appropriate documentation (again: not as a criticism of you or the other devs, but rather of the current process). I added a comment about this in #28568.

**Comment by TomAugspurger** [comment]
@h-vetinari The calls and meeting notes are public.

> to deprecate set_index entirely

Not quite: I was just choosing different names to highlight the different behavior. Clearly set_index should stay :)

**Comment by toobaz** [comment]
> Not quite: I was just choosing different names to highlight the different behavior. Clearly set_index should stay :)

OK, thanks for the clarification ;-)

**Comment by MarkRotchell** [comment]
Can we re-open this discussion for the use case of setting columns based on (row) index labels? `set_index` accepts either column labels or array-like, whereas `set_axis` doesn't accept labels, so can't necessarily be used to set columns based on data in the DataFrame when chaining. For example, consider:

```
>>> df = pd.DataFrame({'a':['foo','bar','baz'], 'b':['spam','eggs','chips']}, index=pd.Index(['red','blue','green'],name='color'))
>>> print(df)
         a      b
color            
red    foo   spam
blue   bar   eggs
green  baz  chips
```

we can chain a `set_index` and use a column label like:

```
>>> df.set_index('a')
         a      b
color            
red    foo   spam
blue   bar   eggs
green  baz  chips
```

but no equivalent for columns, so we can't do:
```
>>> df.set_columns('blue')
blue   bar   eggs
color            
red    foo   spam
green  baz  chips
```

Using `set_axis` with `axis='columns'` requires passing in a list-like or index, which might be fine, e.g.:

```
>>> df.set_axis(df.loc['blue'], axis='columns').drop(index='blue')
blue   bar   eggs
color            
red    foo   spam
green  baz  chips
```
but not if you're chaining, i.e. the below doesn't make sense
```
>>> df.assign(c=['very','naughty','boy']).set_axis(df.loc['blue'], axis='columns').drop(index='blue')
```
It can be achieved with transposition, but this is clunky, and `transpose` requires copies on mixed-dtype arrays

```
>>> df.assign(c=['very','naughty','boy']).transpose().set_index('blue').transpose()
blue   bar   eggs naughty
color                    
red    foo   spam    very
green  baz  chips     boy
```

Obviously it can be done if you're willing to stop chaining and create a new temporary variable, but I always feel annoyed having to break a chain for something that seems so natural for the API to be able to do.

Are there any other ways to do it?

If, not, some possible options to address this use case:

- a new `set_columns` method in the API. 
- adding an `axis` parameter to `set_index`
- allowing `set_axis` to accept labels like `set_index`
